finality-aleph/src/network/address_cache.rs

```diff
@@ -0,0 +1,180 @@
+use std::{collections::HashMap, marker::PhantomData, num::NonZeroUsize, sync::Arc};
+
+use lru::LruCache;
+use parking_lot::Mutex;
+use primitives::{AccountId, AlephSessionApi, AuthorityId, BlockHash, BlockNumber};
+use sc_client_api::Backend;
+use serde::{Deserialize, Serialize};
+use sp_runtime::traits::{Block, Header};
+
+use crate::{
+    abft::NodeIndex,
+    session::{SessionBoundaryInfo, SessionId},
+    session_map::AuthorityProvider,
+    ClientForAleph,
+};
+
+pub trait KeyOwnerInfoProvider {
+    fn aleph_key_owner(&self, block_number: BlockNumber, key: AuthorityId) -> Option<AccountId>;
+}
+
+/// Network details for a given validator. This data is purely informational
+/// and can change over time, even within a single session.
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct ValidatorAddressingInfo {
+    /// Network level address of the validator, i.e. IP address
+    pub network_level_address: String,
+    /// PeerId of the validator used in validator (clique) network
+    pub validator_network_peer_id: String,
+    /// Session to which the given `validator_network_peer_id` corresponds.
+    pub session: SessionId,
+    /// Vec of substrate's P2P network extra data (e.g. peer_id) that *could* match a given addressing info,
+    /// that is, which was sent from the same address as network_level_address.
+    /// It can be empty in case we haven't been ever connected directly to a given peer in p2p network,
+    /// or even could contain false positives when one validator changed its IP and other node picked it up.
+    pub potential_p2p_network_additional_info: Vec<AdditionalP2PAddressingInfo>,
+}
+
+/// Additional information about a node in substrate's P2P network. Use only for debugging purposes,
+/// as content of this struct rely on unstable substrate's API and can change in future.
+#[derive(Debug, Clone, Hash, PartialEq, Eq, Serialize, Deserialize)]
+pub struct AdditionalP2PAddressingInfo {
+    pub p2p_network_peer_id: String,
+    pub version_string: Option<String>,
+}
+
+/// Stores most recent information about validator addresses.
+#[derive(Debug, Clone)]
+pub struct ValidatorAddressCache {
+    data: Arc<Mutex<LruCache<AccountId, ValidatorAddressingInfo>>>,
+}
+
+const VALIDATOR_ADDRESS_CACHE_SIZE: usize = 300;
+
+impl ValidatorAddressCache {
+    pub fn new() -> Self {
+        Self {
+            data: Arc::new(Mutex::new(LruCache::new(
+                NonZeroUsize::try_from(VALIDATOR_ADDRESS_CACHE_SIZE).unwrap(),
+            ))),
+        }
+    }
+
+    pub fn insert(&self, validator_stash: AccountId, info: ValidatorAddressingInfo) {
+        self.data.lock().put(validator_stash, info);
+    }
+
+    pub fn read(&self) -> HashMap<AccountId, ValidatorAddressingInfo> {
```

Comment:
I think it's better to call it `snapshot`, `get_clone` or some such. Otherwise it might be confused with the `read` method from `RwLock` https://amanieu.github.io/parking_lot/parking_lot/struct.RwLock.html#method.read

Reply:
Renamed to `snapshot` and cloning.
> returning readonly reference

Not sure if I understand this, but I'm not aware of any way of converting `&LruCache<>` to `&Hashmap<>`